flocker/control/_persistence.py

```diff
@@ -163,15 +252,66 @@ def __init__(self, reactor, path):
     def startService(self):
         if not self._path.exists():
             self._path.makedirs()
-        self._config_path = self._path.child(b"current_configuration.v1.json")
-        if self._config_path.exists():
+        self.load_configuration()
+        MultiService.startService(self)
+        _LOG_STARTUP(configuration=self.get()).write(self.logger)
+
+    def _versioned_config(self, force_latest=False):
+        """
+        Sequentially and decrementally check for versioned configuration
+        files, from the current version to version 1. Return the file
+        found along with its version number, or if no valid
+        configuration file exists, return the current version number
+        and the expected file path of a configuration file matching
+        the latest version.
+
+        :return: A ``tuple`` comprising a version ``int`` and
+            config ``FilePath``.
+        """
+        config_files = [
+            (
+                version,
+                self._path.child(_VERSIONED_CONFIG_FILE % version)
+            )
+            for version in range(_CURRENT_VERSION, -1, -1)
+        ]
+        # Check for each possible versioned config file, from newest
+        # to oldest and return the first match.
+        for version, config_file in config_files:
```

Comment:
This method could be improved. Right now it does two distinct things:
1. Find latest existing file.
2. Give file name in case config doesn't exist.

These are not really the same thing; notice how you're duplicating a bunch of code. So probably you want two separate functions for each use case.